src/utils/src/ExponentialBackoffUtils.c

```diff
@@ -0,0 +1,222 @@
+    } else {
+        if (pExponentialBackoffState != NULL) {
+            SAFE_MEMFREE(pExponentialBackoffState);
+        }
+    }
+
+    LEAVES();
+    return retStatus;
+}
+
+STATUS validateExponentialBackoffConfig(PExponentialBackoffConfig pBackoffConfig) {
+    ENTERS();
+    STATUS retStatus = STATUS_SUCCESS;
+
+    CHK(pBackoffConfig != NULL, STATUS_NULL_ARG);
+
+    // The upper bound on wait time should be at least DEFAULT_KVS_MAX_WAIT_TIME_MILLISECONDS
+    CHK(pBackoffConfig->maxWaitTime >= DEFAULT_KVS_MAX_WAIT_TIME_MILLISECONDS, STATUS_INVALID_ARG);
+    // The retry factor time should be at least DEFAULT_KVS_RETRY_TIME_FACTOR_MILLISECONDS
+    CHK(pBackoffConfig->retryFactorTime >= DEFAULT_KVS_RETRY_TIME_FACTOR_MILLISECONDS, STATUS_INVALID_ARG);
+    // Minimum time between two consecutive calls for exponential wait before the state is
+    // reset to start from count 0 must be at least DEFAULT_KVS_MIN_TIME_TO_RESET_RETRY_STATE_MILLISECONDS
+    CHK(pBackoffConfig->minTimeToResetRetryState >= DEFAULT_KVS_MIN_TIME_TO_RESET_RETRY_STATE_MILLISECONDS, STATUS_INVALID_ARG);
+    // Jitter factor must be at least DEFAULT_KVS_JITTER_FACTOR_MILLISECONDS
+    CHK(pBackoffConfig->jitterFactor >= DEFAULT_KVS_JITTER_FACTOR_MILLISECONDS, STATUS_INVALID_ARG);
+
+CleanUp:
+    LEAVES();
+    return retStatus;
+}
+
+STATUS exponentialBackoffStateCreate(PExponentialBackoffState* ppBackoffState, PExponentialBackoffConfig pBackoffConfig) {
+    ENTERS();
+    STATUS retStatus = STATUS_SUCCESS;
+    PExponentialBackoffState pExponentialBackoffState = NULL;
+
+    CHK(ppBackoffState != NULL && pBackoffConfig != NULL, STATUS_NULL_ARG);
+    CHK_STATUS(validateExponentialBackoffConfig(pBackoffConfig));
+
+    pExponentialBackoffState = (PExponentialBackoffState) MEMCALLOC(0x00, SIZEOF(ExponentialBackoffState));
+    CHK(pExponentialBackoffState != NULL, STATUS_NOT_ENOUGH_MEMORY);
+
+    pExponentialBackoffState->exponentialBackoffConfig = *pBackoffConfig;
+    CHK_STATUS(resetExponentialBackoffRetryState(pExponentialBackoffState));
+
+CleanUp:
+    if (STATUS_SUCCEEDED(retStatus)) {
+        (*ppBackoffState) = pExponentialBackoffState;
+    } else {
+        SAFE_MEMFREE(pExponentialBackoffState);
+    }
+
+    LEAVES();
+    return retStatus;
+}
+
+UINT64 getRandomJitter(UINT32 jitterFactor) {
+    return (RAND() % jitterFactor) * HUNDREDS_OF_NANOS_IN_A_MILLISECOND;
+}
+
+UINT64 calculateWaitTime(PExponentialBackoffState pRetryState, PExponentialBackoffConfig pRetryConfig) {
+    // DEFAULT_KVS_EXPONENTIAL_FACTOR is 2 and if currentRetryCount > 64, then
+    // power(2, currentRetryCount) will overflow UINT64 buffer. In such case we
+    // always return the MAX_UINT64 value.
+    // This case will occur only if the exponential retries are configured with no
+    // upper bound on wait time. The implicit upper bound in that case will be MAX_UINT64.
+    if (pRetryState->currentRetryCount > 64) {
+        return MAX_UINT64;
+    }
```

Comment:
I mean does it make any sense to do this?  What is being returned is a time this will be some insanely large time.  Also IF you did hit this condition and you returned MAX_UINT64, you WILL overflow down below:

`pRetryState->lastRetryWaitTime = currentTime + waitTime + jitter;`

I think we should enforce sane values here.

Reply:
+1

Reply:
I have added some upper limits for the config parameters. Those will ensure that we never reach insanely high wait time value (which is impractical too). We can get rid of `this check` after those validations :)